hyppo/ksample/smoothCF.py

```diff
@@ -0,0 +1,155 @@
+    :math: '1 - \alpha' quantil of a Chi-squared distribution w/ J degrees of freedom
+    is chosen. Null is rejected if :math: 'S_n' is larger than this threshold.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    def __init__(self, scale=1.0, num_random_features=5, **kwargs):
+
+        self.scale = scale
+        self.num_random_features = num_random_features
+        KSampleTest.__init__(
+            self,  **kwargs
+        )
+
+    def statistic(self, x, y):
+        """
+        :return: test statistic for smoothCF
+        """
+        _, dim_x = np.shape(x)
+        random_frequencies = _gen_random(dim_x, self.num_random_features)
+        difference = _smooth_difference(random_frequencies, self.scale * x, self.scale * y)
+        return _distance(difference, 2 * self.num_random_features)
+
+    def test(self, x, y, random_state=None):
+        r"""
+        Calculates the smooth CF test statistic and p-value.
+
+        Parameters
+        ----------
+        x,y : ndarray
+            Input data matrices. ``x`` and ``y`` must have the same number of
+            dimensions. That is, the shapes must be ``(n, p)`` and ``(m, p)`` where
+            `n` is the number of samples and `p` and `q` are the number of
+            dimensions.
+
+        Returns
+        -------
+        stat : float
+            The computed Smooth CF statistic.
+        pvalue : float
+            The computed smooth CF p-value.
+        """
+        check_input = _CheckInputs(
+            inputs=[x,y],
+            indep_test=None
+        )
+        x, y = check_input()
+
+        stat = self.statistic(x,y)
+        pvalue = chi2.sf(stat, self.num_random_features)
+        self.stat = stat
+        self.pvalue = pvalue
+
+        return KSampleTestOutput(stat, pvalue)
+
+
+@jit(nopython=True, cache=True)
+def _gen_random(dimension, num_random_features):
+    """Generates test points for vector of differences"""
+    return np.random.randn(dimension, num_random_features)
+
+
+@jit(nopython=True, cache=True)
+def _smooth(data):
+    """Smooth kernel"""
+    norms = np.zeros(data.shape[0])
+    for i in range(data.shape[0]):
+        norms[i] = np.sqrt(np.sum(data[i]**2))
+    w = norms
+    w = np.exp(-w ** 2 / 2)
+    return w.reshape(-1,1)
+
+
+@jit(nopython=True, cache=True)
+def _smooth_cf(data, w, random_frequencies):
+    """Vector of differences for Smooth CF"""
+    n, _ = data.shape
+    _, d = random_frequencies.shape
+    mat = data.dot(random_frequencies)
+    arr = np.concatenate((np.sin(mat) * w, np.cos(mat) * w), 1)
+    return arr
+
+
+@jit(nopython=True, cache=True)
+def _smooth_difference(random_frequencies, X, Y):
+    """Vector of differences for Smooth CF"""
+    x_smooth = _smooth(X)
+    y_smooth = _smooth(Y)
+    return _smooth_cf(X, x_smooth, random_frequencies) - _smooth_cf(Y, y_smooth, random_frequencies)
+
+
+def _distance(difference, num_random_features):
```

Comment:
not sure, but can you use things from the Hotelling function already in hyppo

Reply:
added distance as a helper

Reply:
not entirely the same, changed to helper